sonar-ballerina-plugin/src/main/java/io/ballerina/sonar/BallerinaSensor.java

```diff
@@ -0,0 +1,254 @@
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import org.apache.commons.lang3.SystemUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.api.batch.fs.FilePredicate;
+import org.sonar.api.batch.fs.FileSystem;
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.batch.rule.Severity;
+import org.sonar.api.batch.sensor.Sensor;
+import org.sonar.api.batch.sensor.SensorContext;
+import org.sonar.api.batch.sensor.SensorDescriptor;
+import org.sonar.api.rule.RuleKey;
+import org.sonar.api.rules.RuleType;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static io.ballerina.sonar.Constants.ANALYSIS_RESULTS_FILE_PATH;
+import static io.ballerina.sonar.Constants.BUG;
+import static io.ballerina.sonar.Constants.BUILT_IN;
+import static io.ballerina.sonar.Constants.EXTERNAL;
+import static io.ballerina.sonar.Constants.ISSUES_FILE_PATH;
+import static io.ballerina.sonar.Constants.ISSUE_END_LINE;
+import static io.ballerina.sonar.Constants.ISSUE_END_LINE_OFFSET;
+import static io.ballerina.sonar.Constants.ISSUE_FILE_PATH;
+import static io.ballerina.sonar.Constants.ISSUE_MESSAGE;
+import static io.ballerina.sonar.Constants.ISSUE_RULE_ID;
+import static io.ballerina.sonar.Constants.ISSUE_RULE_KIND;
+import static io.ballerina.sonar.Constants.ISSUE_SOURCE;
+import static io.ballerina.sonar.Constants.ISSUE_START_LINE;
+import static io.ballerina.sonar.Constants.ISSUE_START_LINE_OFFSET;
+import static io.ballerina.sonar.Constants.RULE_REPOSITORY_KEY;
+import static io.ballerina.sonar.Constants.SONAR_SCANNER_OFFSET;
+import static io.ballerina.sonar.Constants.VULNERABILITY;
+
+/**
+ * Represents the implementation of the {@link Sensor} class for Ballerina.
+ *
+ * @since 0.1.0
+ */
+class BallerinaSensor implements Sensor {
+    private final Logger logger = LoggerFactory.getLogger(BallerinaSensor.class);
+    private final BallerinaLanguage language;
+    private final Set<String> externalRules = new HashSet<>();
+
+    public BallerinaSensor(BallerinaLanguage language) {
+        this.language = language;
+    }
+
+    @Override
+    public void describe(SensorDescriptor descriptor) {
+        descriptor.onlyOnLanguage(language.getKey()).name(language.getName() + " Sensor");
+    }
+
+    @Override
+    public void execute(SensorContext sensorContext) {
+        FileSystem fileSystem = sensorContext.fileSystem();
+        FilePredicate mainFilePredicate = sensorContext.fileSystem().predicates()
+                .and(fileSystem.predicates().hasLanguage(language.getKey()),
+                        fileSystem.predicates().hasType(InputFile.Type.MAIN));
```

Comment:
Why do we need this?

Reply:
This is used to filter and register all the source files that the scan should be looking at. It accesses the available files and filter using the `BallerinaLanguage.getFileSuffixes` method. 

However, we don't need the `fileSystem.predicates().hasType(InputFile.Type.MAIN)` part. We don't have to filter out just the source files and exclude test files. This doesn't seem to anyway work because even with this condition, the test files are shown in the sonarqube UI with any diagnostics available.